src/sagemaker/local/data.py

```diff
@@ -0,0 +1,249 @@
+from sagemaker.local.utils import download_folder
+from sagemaker.utils import get_config_value
+
+
+class DataSourceFactory(object):
+
+    @staticmethod
+    def get_instance(data_source, sagemaker_session):
+        parsed_uri = urlparse(data_source)
+        if parsed_uri.scheme == 'file':
+            return LocalFileDataSource(parsed_uri.path)
+        elif parsed_uri.scheme == 's3':
+            return S3DataSource(parsed_uri.netloc, parsed_uri.path, sagemaker_session)
+
+
+class DataSource(object):
+
+    def get_file_list(self):
+        pass
+
+    def get_root_dir(self):
+        pass
+
+
+class LocalFileDataSource(DataSource):
+    """
+    Represents a data source within the local filesystem.
+    """
+
+    def __init__(self, root_path):
+        self.root_path = os.path.abspath(root_path)
+        if not os.path.exists(self.root_path):
+            raise RuntimeError('Invalid data source: %s Does not exist.' % self.root_path)
+
+    def get_file_list(self):
+        """Retrieve the list of absolute paths to all the files in this data source.
+
+        Returns:
+             List[string] List of absolute paths.
+        """
+        if os.path.isdir(self.root_path):
+            files = [os.path.join(self.root_path, f) for f in os.listdir(self.root_path)
+                     if os.path.isfile(os.path.join(self.root_path, f))]
+        else:
+            files = [self.root_path]
+
+        return files
+
+    def get_root_dir(self):
+        """Retrieve the absolute path to the root directory of this data source.
+
+        Returns:
+            string: absolute path to the root directory of this data source.
+        """
+        if os.path.isdir(self.root_path):
+            return self.root_path
+        else:
+            return os.path.dirname(self.root_path)
+
+
+class S3DataSource(DataSource):
+    """Defines a data source given by a bucket and s3 prefix. The contents will be downloaded
+    and then processed as local data.
+    """
+
+    def __init__(self, bucket, prefix, sagemaker_session):
+        """Create an S3DataSource instance
+
+        Args:
+            bucket (str): s3 bucket name
+            prefix (str): s3 prefix path to the data
+            sagemaker_session (sagemaker.Session): a sagemaker_session with the desired settings to talk to s3
+
+        """
+
+        # Create a temporary dir to store the S3 contents
+        root_dir = get_config_value('local.container_root', sagemaker_session.config)
+        if root_dir:
+            root_dir = os.path.abspath(root_dir)
+
+        working_dir = tempfile.mkdtemp(dir=root_dir)
+        download_folder(bucket, prefix, working_dir, sagemaker_session)
+        self.files = LocalFileDataSource(working_dir)
+
+    def get_file_list(self):
+        return self.files.get_file_list()
+
+    def get_root_dir(self):
+        return self.files.get_root_dir()
+
+
+class SplitterFactory(object):
+
+    @staticmethod
+    def get_instance(split_type):
```

Comment:
nit: The _get_ prefix does match a factory method. If we are following the Google Python Guide, we should avoid static methods and have a module level function instead https://github.com/google/styleguide/blob/gh-pages/pyguide.md#2174-decision